RecoEgamma/EgammaTools/interface/MultiToken.h

```diff
@@ -0,0 +1,127 @@
+#ifndef RecoEgamma_EgammaTools_MultiToken_H
+#define RecoEgamma_EgammaTools_MultiToken_H
+
+#include "FWCore/Framework/interface/ConsumesCollector.h"
+#include "FWCore/Utilities/interface/EDGetToken.h"
+
+/*
+ * This class is a wrapper around a vector of EDM tokens, of which at least one
+ * is expected to yield a valid handle.
+ *
+ * The first time you call getValidHandle() or getHandle(), it will go over all
+ * tokens and try to get a valid handle. If no token yields a valid handle, it
+ * will either throw and exception or return the last non-valid handle.
+ *
+ * Once it found a valid handle, it will remember which token was used to get
+ * it and therefore doesn't need to loop over all tokens from there on.
+ *
+ * Example use case: auto-detection of AOD vs. MiniAOD.
+ *
+ * Created by Jonas Rembser on August 3, 2018.
+ */
+
+class MultiToken {
+
+  public:
+
+    // Constructor from an arbitrary number of tokens
+    template <typename ... Tokens>
+    MultiToken(edm::EDGetToken token0, Tokens ... tokens)
+      : tokens_( { token0, tokens... } )
+      , goodIndex_(-1)
+    {}
+
+    // Get a handle on the event data, non-valid handle is allowed
+    template <class T>
+    edm::Handle<T> getHandle(const edm::Event& iEvent)
+    {
+        edm::Handle<T> handle;
+
+        // If we already know which token works, take that one
+        if (goodIndex_ >= 0) {
+            iEvent.getByToken(tokens_[goodIndex_], handle);
+            return handle;
+        }
+
+        // If not, set the good token index parallel to getting the handle
+        handle = getInitialHandle<T>(iEvent);
+
+        if (goodIndex_ == -1) {
+            goodIndex_ = tokens_.size() - 1;
+        }
+        return handle;
+    }
+
+    // Get a handle on the event data,
+    // throw exception if no token yields a valid handle
+    template <class T>
+    edm::Handle<T> getValidHandle(const edm::Event& iEvent)
+    {
+        edm::Handle<T> handle;
+
+        // If we already know which token works, take that one
+        if (goodIndex_ >= 0) {
+            iEvent.getByToken(tokens_[goodIndex_], handle);
+            if (!handle.isValid())
+                throw cms::Exception("MultiTokenException") <<
+                    "Token gave valid handle in previous event but not anymore!";
+            return handle;
+        }
+
+        // If not, set the good token index parallel to getting the handle
+        handle = getInitialHandle<T>(iEvent);
+
+        if (goodIndex_ == -1) {
+            throw cms::Exception("MultiTokenException") << "Neither handle is valid!";
+        }
+        return handle;
+    }
+
+    // get the good token
+    template <class T>
+    edm::EDGetToken get(const edm::Event& iEvent)
```

Comment:
in line with https://github.com/cms-sw/cmssw/pull/24312#issuecomment-415523802
it may help to add an index to the argument list
so that it is possible to enforce what we need to get